Sources/ImagePrefetcher.swift

```diff
@@ -0,0 +1,119 @@
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+
+#if os(OSX)
+    import AppKit
+#else
+    import UIKit
+#endif
+
+
+/// Progress update block of prefetcher.
+public typealias PrefetchProgressBlock = ((completedURLs: Int, allURLs: Int) -> ())
+
+/// Completion block of prefetcher.
+public typealias PrefetchCompletionBlock = ((completedURLs: Int, skippedURLs: Int) -> ())
+
+
+/// `ImagePrefetcher` represents a downloading manager for requesting many images via URLs and caching them.
+public class ImagePrefetcher: NSObject {
+    
+    private var prefetchURLs: [NSURL]?
+    private var skippedCount = 0
+    private var requestedCount = 0
+    private var finishedCount = 0
+    
+    private var downloader: ImageDownloader
+
+    /// The maximum concurrent downloads to use when prefetching images. Default is 5.
+    var maxConcurrentDownloads = 5
+    
+    public init(downloader: ImageDownloader) {
+        self.downloader = downloader
+        super.init()
+    }
+    
+    /**
+     Download the images from `urls` and cache them. This can be useful for background downloading
+     of assets that are required for later use in an app. This code will not try and update any UI
+     with the results of the process, but calls the handlers with the number cached etc. Failed
+     images are just skipped.
+     
+     - parameter urls:              The list of URLs to prefetch
+     - parameter progressBlock:     Block to be called when progress updates. Completed and total
+                                    counts are provided. Completed does not imply success.
+     - parameter completionHandler: Block to be called when prefetching is complete. Completed is all
+                                    those made, and skipped is the number of failed ones.
+     */
+    public func prefetchURLs(urls: [NSURL], progressBlock: PrefetchProgressBlock?, completionHandler: PrefetchCompletionBlock?) {
+        
+        // Clear out any existing prefetch operation first
+        cancelPrefetching()
+        
+        prefetchURLs = urls
+        
+        guard urls.count > 0 else {
+            CompletionHandler?()
+            return
+        }
+        
+        for (var i = 0; i < maxConcurrentDownloads && requestedCount < urls.count; i++) {
+            startPrefetching(i, progressBlock: progressBlock, completionHandler: completionHandler)
+        }
+    }
+   
+    internal func cancelPrefetching() {
```

Comment:
The cancel logic contains some gutter here. Consider about the situation that you call `prefetchURLs` before the previous `prefetchURLs` download finishing.

Because there is no cancelling now for the download process, there is a chance that the `completionHandler` of `downloadAndCacheImageWithURL` for previous fetch is called wrongly and `finishedCount` gets added by mistake.